Revise Observability section (#296)
kgateway-dev/kgateway.dev: content/docs/observability/otel-stack.md

    D --> D1["Logs (Loki)"]
    D --> D2["Traces (Tempo)"]
    D --> D3["Metrics (Prometheus)"]
    D1 --> E["5- Visualization (Grafana)"]
    D2 --> E
    D3 --> E
```
Architecture data flow:
1. **Application Traffic**: Applications send requests to the gateway proxy.
2. **Gateway Processing**: The gateway proxy processes requests and generates telemetry data in the form of logs, traces, and metrics.
3. **Telemetry Collection**: The OTel Collectors receive telemetry data from the gateway proxy.
4. **Data Storage**: The OTel Collectors route data to the appropriate storage backends:
   - **Logs** go to Loki for log aggregation and storage.
   - **Traces** go to Tempo for distributed tracing storage.
   - **Metrics** go to Prometheus for time-series metrics storage.
5. **Visualization**: Grafana queries the storage backends as data sources to create unified dashboards.

## Before you begin

{{< reuse "docs/snippets/prereq.md" >}}

## Step 1: Install Grafana Loki and Tempo {#grafana}

Grafana is a suite of open source tools that help you analyze, visualize, and monitor data in your cluster. For the OTel stack, you install the following Grafana components:

> * **Loki**: A log aggregation system that indexes only the metadata about your logs (_labels_) and not the actual log contents. This way, Loki is more cost-efficient and performant than traditional log aggregation systems.

Reviewer: This is what Mr. AI used as a reference probably:
>  Unlike other logging systems, Loki does not index the contents of the logs, but only indexes metadata about your logs as a set of labels for each log stream.

Reference: https://grafana.com/docs/loki/latest/get-started/overview/

> * **Loki**: A log aggregation system that indexes only the metadata about your logs (_labels_) and not the actual log contents. This way, Loki is more cost-efficient and performant than traditional log aggregation systems.

Author: the labels are what it indexes, will rephrase

       namespace: {{< reuse "docs/snippets/namespace.md" >}}
     to:
     - group: ""
       kind: Service
       name: opentelemetry-collector-traces
   EOF
   ```

## Step 5: Verify your setup {#verify}

To verify that your setup is working, generate sample traffic and review the logs and Grafana dashboard.
   
1. Generate traffic for the httpbin app. 

   {{< tabs items="Cloud Provider LoadBalancer,Port-forward for local testing" tabTotal="2">}}
   {{% tab tabName="Cloud Provider LoadBalancer" %}}
   ```sh
   for i in {1..5}; do curl -v http://$INGRESS_GW_ADDRESS:8080/status/418 -H "host: www.example.com:8080"; done
   ```
   {{% /tab %}}
   {{% tab tabName="Port-forward for local testing" %}} %}} %}}
   ```sh
   for i in {1..5}; do curl -v localhost:8080/status/418 -H "host: www.example.com:8080"; done
   ```
   {{% /tab %}}
   {{< /tabs >}}

   Example output:

   ```
   < HTTP/1.1 418 Unknown
   I'm a teapot!
   ...
   ```

2. Check that logs are being collected.

   ```sh
   kubectl -n telemetry logs deploy/opentelemetry-collector-logs | grep '/status/418' | wc -l
   ```

   Example output: The count matches the number of requests that you sent, such as `5`.

   ```

> 

Reviewer 2: assuming these blank lines are in the ex output intentionally, ignore if so

Author: yeah, it is how it rendered in mine, but maybe not needed? it was just weird, so thought I'd include:

```
arthurberger@artberger-MacBook-Pro ~ % kubectl -n telemetry logs deploy/opentelemetry-collector-logs | grep '/status/418' | wc -l

       5
arthurberger@artberger-MacBook-Pro ~ % kubectl -n telemetry logs deploy/opentelemetry-collector-traces | grep 'http.status_code: Str(418)' | wc -l

      10
```

       5
   ```

3. Check that traces are being collected.

   ```sh
   kubectl -n telemetry logs deploy/opentelemetry-collector-traces | grep 'http.status_code: Str(418)' | wc -l
   ```

   Example output: The count traces the number of services that were involved in responding to the request, such as `10`.

   ```

      10
   ```

3. Verify that data plane metrics for the gateway proxy are being collected by using a Grafana dashboard.
   
   1. Create a Grafana dashboard for the data plane metrics of your gateway proxy. For example, you can download the [sample Grafana dashboard configuration for the `http` gateway](grafana.json) as an `envoy.json` file. 

   2. Import the Grafana dashboard.
   
      ```sh
      kubectl -n telemetry create cm envoy-dashboard \
      --from-file=envoy.json
      kubectl label -n telemetry cm envoy-dashboard grafana_dashboard=1
      ```

>    3. Open Grafana and log in to Grafana by using the username `admin` and password `prom-operator`. 

Reviewer 2: ```suggestion
   3. Open Grafana and log in by using the username `admin` and password `prom-operator`. 
```

      
      {{< tabs items="Cloud Provider LoadBalancer,Port-forward for local testing" tabTotal="2">}}
      {{% tab tabName="Cloud Provider LoadBalancer" %}}
      ```sh
      open "http://$(kubectl -n telemetry get svc kube-prometheus-stack-grafana -o jsonpath="{.status.loadBalancer.ingress[0]['hostname','ip']}"):3000"
      ```
      {{% /tab %}}

>       {{% tab tabName="Port-forward for local testing" %}} %}} %}}

Reviewer 2: ```suggestion
      {{% tab tabName="Port-forward for local testing" %}}
```

   - **Logs** go to Loki for log aggregation and storage.
   - **Traces** go to Tempo for distributed tracing storage.
   - **Metrics** go to Prometheus for time-series metrics storage.
5. **Visualization**: Grafana queries the storage backends as data sources to create unified dashboards.

### More considerations {#more-considerations}

**Push model**: This guide sets up the OTel collectors to push metrics to the storage backends (`push` model), vs. setting up the backends such as Prometheus to scrape metrics from the collector pod (`pull` model). The `push` model is used because it shows the ease and consistency of using OTel for demonstration purposes. It also supports Native Histograms out of the box, which the `pull` model does not due to [a known OTel issue with the Prometheus exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33703).

**Debug exporter**: The example pipelines in all three OTel collectors set up the `debug` exporter. This exporter is useful for testing and validation purposes. However, for production scenarios, remove this exporter to avoid performance impacts.

> **Prometheus exporter**: You can use the `promexporter` endpoint with Prometheus to scrape metrics from the collector pod, if you prefer the `pull` model to the `push` model. Also, if you use the `pull` model, make sure to configure Prometheus to handle Native Histograms per the [known issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33703) previously mentioned.

Reviewer: ```suggestion
**Prometheus exporter**: You can use `prometheusexporter`'s `promexporter` port with Prometheus to scrape metrics from the collector pod (already configured in the example below), if you prefer the `pull` model to the `push` model. Also, if you use the `pull` model, make sure to configure Prometheus to handle Native Histograms and scrape the metrics directly as for this model OTel's `prometheusexporter` is not yet supperted as per the [known issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33703) previously mentioned.
```

By using an OTel collector to aggregate metrics, you avoid having to configure each application individually to send their metrics to each backend observability tool. This setup simplifies your setup, lets you more easily change backends, improves reliability and debuggability, and lets you optimize preprocessing activities such as filtering, transforming, or enriching the metrics before scraping.

You can deploy three separate OTel collectors that are optimized for the three different types of telemetry data: metrics, logs, and traces. This way, you can scale and optimize each collector based on your telemetry needs.

{{< callout type="warning" >}}
The example pipelines in all three OTel collectors set up the `debug` exporter. This exporter is useful for testing and validation purposes. However, for production scenarios, remove this exporter to avoid performance impacts.
{{< /callout >}}

1. Deploy the metrics collector to handle numerical measurements and time-series data. Note that you can also use the `promexporter` endpoint with Prometheus to scrape metrics from the collector pod, if you prefer the `pull` model to the `push` model.

   ```yaml

>    helm upgrade --install opentelemetry-collector-metrics opentelemetry-collector \

Reviewer 2: Not sure if this is important, but got these warnings about resource limits when I installed. Should we set some in the settings here?

```
NOTES:
[WARNING] No resource limits or requests were set. Consider setting resource requests and limits for your collector(s) via the `resources` field.

[WARNING] "useGOMEMLIMIT" is enabled but memory limits have not been supplied so the GOMEMLIMIT env var could not be added. Solve this problem by setting resources.limits.memory or disabling useGOMEMLIMIT
```

Reviewer: I wouldn't bother, otherwise people will take it as granted then complain why our examples are not working at scale.
